File: ios/CryptoSavingsTracker/Utilities/VisualSystemRollout.swift
import Foundation

enum VisualSystemFlow: String, CaseIterable {
    case planning
    case dashboard
    case settings

    var waveKey: String {
        switch self {
        case .planning:
            return VisualSystemRollout.flagWave1Planning
        case .dashboard:
            return VisualSystemRollout.flagWave2Dashboard
        case .settings:
            return VisualSystemRollout.flagWave3Settings
        }
    }

    var waveName: String {
        switch self {
        case .planning:
            return "wave1"
        case .dashboard:
            return "wave2"
        case .settings:
            return "wave3"
        }
    }
}

protocol VisualSystemRemoteConfigProviding {
    func boolValue(for key: String) -> Bool?
}

struct NullVisualSystemRemoteConfigProvider: VisualSystemRemoteConfigProviding {
    func boolValue(for key: String) -> Bool? { nil }
}

protocol VisualSystemTelemetryProviding {
    func track(event: String, payload: [String: String])
}

struct AppLogVisualSystemTelemetryProvider: VisualSystemTelemetryProviding {
    func track(event: String, payload: [String: String]) {
        let ordered = payload
            .sorted(by: { $0.key < $1.key })
            .map { "\($0.key)=\($0.value)" }
            .joined(separator: ",")
        AppLog.info("[\(event)] \(ordered)", category: .ui)
    }
}

final class VisualSystemRollout {
    static let flagWave1Planning = "visual_system.wave1_planning"
    static let flagWave2Dashboard = "visual_system.wave2_dashboard"
    static let flagWave3Settings = "visual_system.wave3_settings"

    private enum Source: String {
        case releaseDefault = "release_default"
        case remoteConfig = "remote_config"
        case debugOverride = "debug_override"
    }

    static let shared = VisualSystemRollout(
        remoteConfigProvider: NullVisualSystemRemoteConfigProvider(),
        telemetryProvider: AppLogVisualSystemTelemetryProvider()
    )

    private let remoteConfigProvider: VisualSystemRemoteConfigProviding
    private let telemetryProvider: VisualSystemTelemetryProviding
    private let userDefaults: UserDefaults
    private let nowProvider: () -> Date
    private var lastEvaluatedValues: [VisualSystemFlow: Bool] = [:]

    init(
        remoteConfigProvider: VisualSystemRemoteConfigProviding,
        telemetryProvider: VisualSystemTelemetryProviding,
        userDefaults: UserDefaults = .standard,
        nowProvider: @escaping () -> Date = { Date() }
    ) {
        self.remoteConfigProvider = remoteConfigProvider
        self.telemetryProvider = telemetryProvider
        self.userDefaults = userDefaults
        self.nowProvider = nowProvider
    }

    func isEnabled(flow: VisualSystemFlow) -> Bool {
        let (value, source) = resolvedValue(for: flow)
        emit(
            event: "vsu_flag_evaluated",
            flow: flow,
            source: source
        )

        let previous = lastEvaluatedValues[flow]
        if previous == true, value == false {
            emit(
                event: "vsu_wave_rollback_triggered",
                flow: flow,
                source: source
            )
            emit(
                event: "vsu_wave_rollback_completed",
                flow: flow,
                source: source
            )
        }
        lastEvaluatedValues[flow] = value
        return value
    }

    func setDebugOverride(_ value: Bool?, for flow: VisualSystemFlow) {
        let key = debugOverrideKey(for: flow)
        if let value {
            userDefaults.set(value, forKey: key)
        } else {
            userDefaults.removeObject(forKey: key)
        }
    }

    private func resolvedValue(for flow: VisualSystemFlow) -> (Bool, Source) {
        let releaseDefault = false
        var value = releaseDefault
        var source: Source = .releaseDefault

        if let remoteValue = remoteConfigProvider.boolValue(for: flow.waveKey) {
            value = remoteValue
            source = .remoteConfig
        }

        if let debugValue = userDefaults.object(forKey: debugOverrideKey(for: flow)) as? Bool {
            value = debugValue
            source = .debugOverride
        }

        return (value, source)
    }

    private func debugOverrideKey(for flow: VisualSystemFlow) -> String {
        "visual_system.debug_override.\(flow.waveKey)"
    }

    private func emit(event: String, flow: VisualSystemFlow, source: Source) {
        let timestamp = ISO8601DateFormatter().string(from: nowProvider())
        telemetryProvider.track(
            event: event,
            payload: [
                "wave": flow.waveName,
                "flow": flow.rawValue,
                "source": source.rawValue,
                "timestamp": timestamp
            ]
        )
    }
}

File: android/app/src/main/java/com/xax/CryptoSavingsTracker/presentation/config/VisualSystemRollout.kt
package com.xax.CryptoSavingsTracker.presentation.config

import android.content.Context
import android.content.SharedPreferences
import android.util.Log

enum class VisualSystemFlow(
    val flowName: String,
    val waveName: String,
    val flagKey: String
) {
    PLANNING(
        flowName = "planning",
        waveName = "wave1",
        flagKey = VisualSystemRollout.FLAG_WAVE1_PLANNING
    ),
    DASHBOARD(
        flowName = "dashboard",
        waveName = "wave2",
        flagKey = VisualSystemRollout.FLAG_WAVE2_DASHBOARD
    ),
    SETTINGS(
        flowName = "settings",
        waveName = "wave3",
        flagKey = VisualSystemRollout.FLAG_WAVE3_SETTINGS
    )
}

interface VisualSystemRemoteConfigAdapter {
    fun boolValue(key: String): Boolean?
}

class NullVisualSystemRemoteConfigAdapter : VisualSystemRemoteConfigAdapter {
    override fun boolValue(key: String): Boolean? = null
}

interface VisualSystemTelemetrySink {
    fun track(event: String, payload: Map<String, String>)
}

class LogcatVisualSystemTelemetrySink : VisualSystemTelemetrySink {
    override fun track(event: String, payload: Map<String, String>) {
        val serialized = payload.toSortedMap().entries.joinToString(",") { "${it.key}=${it.value}" }
        Log.i("VisualSystemRollout", "[$event] $serialized")
    }
}

class VisualSystemRollout(
    private val remoteConfigAdapter: VisualSystemRemoteConfigAdapter,
    private val telemetrySink: VisualSystemTelemetrySink,
    private val prefs: SharedPreferences,
    private val nowMillis: () -> Long = { System.currentTimeMillis() }
) {
    companion object {
        const val FLAG_WAVE1_PLANNING = "visual_system.wave1_planning"
        const val FLAG_WAVE2_DASHBOARD = "visual_system.wave2_dashboard"
        const val FLAG_WAVE3_SETTINGS = "visual_system.wave3_settings"
        private const val PREFS_NAME = "visual_system_rollout"
        private const val DEBUG_PREFIX = "visual_system.debug_override."

        fun from(
            context: Context,
            remoteConfigAdapter: VisualSystemRemoteConfigAdapter = NullVisualSystemRemoteConfigAdapter(),
            telemetrySink: VisualSystemTelemetrySink = LogcatVisualSystemTelemetrySink()
        ): VisualSystemRollout {
            return VisualSystemRollout(
                remoteConfigAdapter = remoteConfigAdapter,
                telemetrySink = telemetrySink,
                prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            )
        }
    }

    private enum class Source(val wireName: String) {
        RELEASE_DEFAULT("release_default"),
        REMOTE_CONFIG("remote_config"),
        DEBUG_OVERRIDE("debug_override")
    }

    private val releaseDefaults: Map<VisualSystemFlow, Boolean> = mapOf(
        VisualSystemFlow.PLANNING to false,
        VisualSystemFlow.DASHBOARD to false,
        VisualSystemFlow.SETTINGS to false
    )
    private val lastEvaluatedValues: MutableMap<VisualSystemFlow, Boolean> = mutableMapOf()

    fun isEnabled(flow: VisualSystemFlow): Boolean {
        val (value, source) = resolveValue(flow)
        emit("vsu_flag_evaluated", flow, source)

        val previous = lastEvaluatedValues[flow]
        if (previous == true && !value) {
            emit("vsu_wave_rollback_triggered", flow, source)
            emit("vsu_wave_rollback_completed", flow, source)
        }
        lastEvaluatedValues[flow] = value
        return value
    }

    fun setDebugOverride(flow: VisualSystemFlow, value: Boolean?) {
        val key = debugKey(flow)
        prefs.edit().apply {
            if (value == null) remove(key) else putBoolean(key, value)
        }.apply()
    }

    private fun resolveValue(flow: VisualSystemFlow): Pair<Boolean, Source> {
        var value = releaseDefaults[flow] ?: false
        var source = Source.RELEASE_DEFAULT

        remoteConfigAdapter.boolValue(flow.flagKey)?.let {
            value = it
            source = Source.REMOTE_CONFIG
        }

        if (prefs.contains(debugKey(flow))) {
            value = prefs.getBoolean(debugKey(flow), value)
            source = Source.DEBUG_OVERRIDE
        }
        return value to source
    }

    private fun debugKey(flow: VisualSystemFlow): String = "$DEBUG_PREFIX${flow.flagKey}"

    private fun emit(event: String, flow: VisualSystemFlow, source: Source) {
        telemetrySink.track(
            event = event,
            payload = mapOf(
                "wave" to flow.waveName,
                "flow" to flow.flowName,
                "source" to source.wireName,
                "timestamp" to nowMillis().toString()
            )
        )
    }
}
